### DISCUSSION TITLE
Training with a set of ungeoreferenced image files

### DISCUSSION THREAD

**Discussion by MathiasBaumgartinger** [azavea/raster-vision#1585] [Q&A]
Hello!

I am using a dataset that consists of individual ungeoreferenced image files for training (`.bmp` in my specific case)? As suggested in the docs:
> If \> 1 URI is provided, a VRT will be created to mosaic together the individual images.

Consequently, I get `Warning 1: gdalbuildvrt does not support ungeoreferenced image. Skipping /path/to/image.bmp` and the subsequent error `rasterio.errors.RasterioIOError: /tmp/tmp7z6zermj/tmphkqhto1f/tmprocf63op/index.vrt: No such file or directory`.

Is there any possibility to circumvent this problem for training? I am not using a plain pytorch model because the predictions shall be performed on large tiffs again.

**Comment by AdeelH** [comment]
- What version/branch/docker-image of Raster Vision are you using?
- Are these `.bmp` files small chips/patches or large scenes?
- Raster Vision can do predictions on large TIFFs just fine. What is your exact use case?

**Reply by MathiasBaumgartinger** [reply]
Thanks for the quick answer - much appreciated!

- I am using rastervision 0.13 natively on Manjaro Linux
- The `.bmp`s are small chips
- I am aware that the prediction on TIFF is working. I just wanted to clarify why I want to use rastervision as opposed to a plain pytorch model (which would work fine on chipped image files like these). I want to train on chips and later predict tiff scenes - the classes to be segmented are grounded and rooftop photovoltaic panels.

**Reply by AdeelH** [reply]
I can think of a few ways to go about this:

## Option 1
Here is some sample code for accomplishing this with 0.13: https://github.com/azavea/raster-vision/issues/1392#issuecomment-1108556397

Edit: actually, even this might not be entirely 0.13-compatible.

## Option 2
However, I would strongly recommend using the master branch or waiting a few days until the new version (0.20) is released.

If you use the latest version, you can train a model outside of RV and only use RV for handling the GeoTIFF-reading and prediction-writing part.

Take a look at this tutorial notebook for an example: https://raster-vision--1586.org.readthedocs.build/en/1586/usage/tutorials/lightning_workflow.html. Note how `val_ds` and `val_dl` are defined -- that's the GeoTIFF-reading part. Then see the _Make predictions for scene_ and the _Visualize and then save predictions_ sections in the same tutorial.

## Option 3

If using the latest version, you can do the same thing as option 1 while using RV as a normal library instead of first defining a `get_config()` function and then running it from the command line.

Here is a tutorial on how to do that: https://raster-vision--1586.org.readthedocs.build/en/1586/usage/tutorials/train.html . Note that the tutorial is for training from GeoTIFFs, so in your case you will have to modify the code to use [SemanticSegmentationImageDataset](https://raster-vision--1586.org.readthedocs.build/en/1586/api_reference/_generated/rastervision.pytorch_learner.dataset.semantic_segmentation_dataset.SemanticSegmentationDataReader.html#rastervision.pytorch_learner.dataset.semantic_segmentation_dataset.SemanticSegmentationDataReader) instead.

---

Let me know if you have any questions.

**Reply by MathiasBaumgartinger** [reply]
Hello @AdeelH again...

I have managed to get things going with the option 3. The training process runs and I was able to save multiple different models via the `save_model_bundle()` function. However, when trying to predict new imagery I have stumbled across some problems:

First of all, I want to clarify that I am using the model which is described in the training tutorial:  
```py
model = torch.hub.load(
    'AdeelH/pytorch-fpn',
    'make_fpn_resnet',
    name='resnet18',
    fpn_type='panoptic',
    ...
)
```

However, the field `model` in the resulting `pipeline_config.json` after saving the trained model will be null. This causes an error when calling the function `learner = SemanticSegmentationLearner.from_model_bundle(bundle_uri, training=False)`. Since the model is available online, I figured i could try a workaround; instead of loading the model via the model-bundle, I tried loading the `pipeline-config.json` and the model weights (`model.pth`) from the model-bundle separately and use the torch hub to load the model as can the code snippet above:

```py
config_dict = file_to_json(model_base_uri + "pipeline-config.json")
config_dict = upgrade_config(config_dict)
learner_pipeline_cfg = build_config(config_dict)
cfg = learner_pipeline_cfg.learner

model_weights_path = model_base_uri + "model.pth"
learner = SemanticSegmentationLearner(model=model, cfg=cfg, output_dir="/opt/data/output", 
    model_weights_path=model_weights_path, training=False)
```

This works in theory, but the resulting predictions seem so far off (I tested multiple different learning settings and epochs) that i fear the Resnet is not configured properly. 

Have you got any idea what I am doing wrong?

**Reply by AdeelH** [reply]
> This causes an error when calling the function `learner = SemanticSegmentationLearner.from_model_bundle(bundle_uri, training=False)`.

Can you share the stack trace for this error?

**Reply by MathiasBaumgartinger** [reply]
```
Traceback (most recent call last):
  File "/opt/src/code/rv_predict.py", line 21, in <module>
    learner = SemanticSegmentationLearner.from_model_bundle(bundle_uri, training=False)
  File "/opt/src/rastervision_pytorch_learner/rastervision/pytorch_learner/learner.py", line 998, in from_model_bundle
    cfg.model.pretrained = False
AttributeError: 'NoneType' object has no attribute 'pretrained'
```

Edit: maybe i should add I am running inside a `quay.io/azavea/raster-vision:pytorch-latest`-container

As said earlier, the pipeline-config lacks a model field, which I think causes the error:

```json
{
  "root_uri": null,
  "rv_config": {},
  "plugin_versions": {
    "rastervision.pytorch_backend": 1,
    "rastervision.pytorch_learner": 4,
    "rastervision.core": 7,
    "rastervision.gdal_vsi": 0,
    "rastervision.aws_batch": 0,
    "rastervision.aws_s3": 0,
    "rastervision.pipeline": 0
  },
  "type_hint": "learner_pipeline",
  "learner": {
    "model": null,
    ...
  ```

**Reply by AdeelH** [reply]
Thank you. Yeah, that's a bug. The `cfg.model.pretrained = False` thing should only happen if `cfg.model` is not `None`.

Looks like you were able to work around this.

>This works in theory, but the resulting predictions seem so far off (I tested multiple different learning settings and epochs) that i fear the Resnet is not configured properly.

You are right. If you pass in a model, RV does not make any changes to it:
https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/learner.py#L341-L359

There should ideally be an error or a warning if a user passes in both a model and a `model_weights_path`.

The solution is to load the weights in yourself before passing it to the `SemanticSegmentationLearner`. You can do that via 

```python
model.load_state_dict(torch.load(model_weights_path))
```

LMK if that works.

**Reply by AdeelH** [reply]
@MathiasBaumgartinger 

We just released v0.20. We have added a new [tutorial for working with non-GeoTIFF datastets](https://docs.rastervision.io/en/0.20/usage/tutorials/prechipped_datasets.html). We have also added [a new section](https://docs.rastervision.io/en/0.20/usage/tutorials/train.html#Using-model-bundles) to the training tutorial about how to use model-bundles; notably, the bug you ran into has been fixed and also passing weights as `model_weights_path` (as you were doing before) should work now.

**Reply by MathiasBaumgartinger** [reply]
Much appreciated, will look into the new version =]

Your workaround with loading the state dict did indeed work! Thanks for your support.
